rakelib/artifacts.rake

```diff
@@ -55,20 +65,76 @@ namespace "artifact" do
     end.flatten.uniq
   end
 
-  task "all" => ["prepare"] do
-    Rake::Task["artifact:deb"].invoke
-    Rake::Task["artifact:rpm"].invoke
-    Rake::Task["artifact:zip"].invoke
-    Rake::Task["artifact:tar"].invoke
+  desc "Generate rpm, deb, tar and zip artifacts"
+  task "all" => ["prepare", "build"]
+
+  desc "Build a tar.gz of default logstash plugins with all dependencies"
+  task "tar" => ["prepare", "generate_build_metadata"] do
+    puts("[artifact:tar] Building tar.gz of default plugins")
+    build_tar
+  end
+
+  desc "Build a zip of default logstash plugins with all dependencies"
+  task "zip" => ["prepare", "generate_build_metadata"] do
+    puts("[artifact:zip] Building zip of default plugins")
+    build_zip
+  end
+
+  desc "Build an RPM of logstash with all dependencies"
+  task "rpm" => ["prepare", "generate_build_metadata"] do
+    puts("[artifact:rpm] building rpm package")
+    package("centos", "5")
+  end
+
+  desc "Build a DEB of logstash with all dependencies"
+  task "deb" => ["prepare", "generate_build_metadata"] do
+    puts("[artifact:deb] building deb package")
+    package("ubuntu", "12.04")
+  end
+
+  desc "Generate logstash core gems"
+  task "gems" => ["prepare"] do
+    Rake::Task["artifact:build-logstash-core"].invoke
+    Rake::Task["artifact:build-logstash-core-event"].invoke
+    Rake::Task["artifact:build-logstash-core-plugin-api"].invoke
+  end
+
+  # "all-plugins" version of tasks
+  desc "Generate rpm, deb, tar and zip artifacts (\"all-plugins\" version)"
+  task "all-all-plugins" => ["prepare-all", "build"]
+
+  desc "Build a zip of all logstash plugins from logstash-plugins github repo"
+  task "zip-all-plugins" => ["prepare-all", "generate_build_metadata"] do
+    puts("[artifact:zip] Building zip of all plugins")
+    build_zip "-all-plugins"
+  end
+
+  desc "Build a tar.gz of all logstash plugins from logstash-plugins github repo"
+  task "tar-all-plugins" => ["prepare-all", "generate_build_metadata"] do
+    puts("[artifact:tar] Building tar.gz of all plugins")
+    build_tar "-all-plugins"
   end
 
-  task "all-all-plugins" => ["prepare-all"] do
+  # Auxiliary tasks
+  task "build" => [:generate_build_metadata] do
+    Rake::Task["artifact:gems"].invoke unless SNAPSHOT_BUILD
     Rake::Task["artifact:deb"].invoke
     Rake::Task["artifact:rpm"].invoke
     Rake::Task["artifact:zip"].invoke
     Rake::Task["artifact:tar"].invoke
   end
 
+  task "generate_build_metadata" do
+    return if defined?(BUILD_METADATA_FILE)
+    BUILD_METADATA_FILE = Tempfile.new('build.rb')
+    build_sha = `git rev-parse HEAD`.chomp
+    metadata = [
+      "# encoding: utf-8",
+      "BUILD_INFO = { \"build_sha\" => \"#{build_sha}\", \"build_snapshot\" => #{SNAPSHOT_BUILD} }",
+    ]
```

Comment:
Small suggestion instead of dealing with manually generating the hash you can use `to_s` to get the ruby code.

```
IO.write("/tmp/test", { "build_sha" => "1234"}.to_s)

~/e/logstash git:6.0 ❯❯❯ cat /tmp/test                                                                                                                  ✱ ◼
{"build_sha"=>"1234"}%
```